### DISCUSSION TITLE
What's the best practice for recurring jobs?

### DISCUSSION THREAD

**Discussion by udoless** [jobrunr/jobrunr#538] [Q&A] [answered]
**Are you running the latest version of JobRunr?**
Yes

**Describe the bug**
I have a long-time(1 ~ 2 hours) running recurring job in k8s.
I deployed my code, and run normally. When I updated my code, stop old container and start new container, the job run on the old containers stopped with `java.lang.IllegalThreadStateException: Job was too long in PROCESSING state without being updated`. But the job was not start on new container on time, I checked jobrunr_jobs table in mysql, scheduledAt field has a few minutes difference from now.

**Environment**
I'm using JobRunr version: 5.1.6
I'm running on JRE / JDK (e.g. OpenJDK 1.8.0_292): OpenJDK 1.8.0_292
I'm using the following StorageProvider (e.g. Oracle / MongoDB / ...):  mysql

**How to Reproduce**

**Expected behavior**
When I update my code, the job should continue be schuduled.

**Additional context**
The job has a constant jobId.

**Comment by udoless** [comment]
I have understand why scheduledAt has a few minutes difference. Because after I updated my code, the retry count will +1, so the schedule deplay will be extended. 
```
protected long getExponentialBackoffPolicy(Job job, int seed) {
        return (long) Math.pow(seed, getFailureCount(job));
}
```

So, what's the best practice for recurring jobs?
Updating code is a normal operation, but it's be treated as exception `java.lang.IllegalThreadStateException: Job was too long in PROCESSING state without being updated`, and it will impact next schedule time.

**Comment by rdehuyss** [comment, marked as answer]
Well, what would you expect to happen?

As the job is still running while the k8s cluster goes down, it cannot finish correctly. On startup JobRunr detects this, fails the jobs (as it is not being processed anymore) and reschedules it.

I'm thinking of adding an option that JobRunr would see whether it missed any recurring jobs and if so, also schedules them, but that will only be available in the Pro version as it's a corner case in my eyes.

Your other option is to wait for processing to complete and only then redeploy.

**Reply by udoless** [reply]
I want to
 stop job by a normal way, because
 in the case above job treated
as failed, lead to retry+1, when retry reach max retry value job will not be scheduled.

**Reply by rdehuyss** [reply]
The job cannot be stopped in a normal way as the k8s cluster is stopped (thus job does not end correctly => fails). I think it's already a good thing that JobRunr will retry automatically after the k8s cluster starts again.

The only thing I can think of is increasing your retry count. In that case, you can configure it to retry up to 20 times for example. If this does not help, I'm afraid I can't help you.
